File: src/checks/meters.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

export const listMetersCheck: Check = {
  name: 'Meters List',
  description: 'List available meters',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const sdk = (client as unknown as {
        sdk: {
          listMeters?: () => Promise<{ data: Array<{ id: string; name: string; unit: string }> } | Array<{ id: string; name: string; unit: string }>>;
        };
      }).sdk;

      if (!sdk.listMeters) {
        const duration = performance.now() - start;
        return {
          name: 'Meters List',
          success: true,
          duration,
          message: 'Skipped (listMeters not available)',
          details: 'The listMeters method is not available in the SDK',
        };
      }

      const result = await sdk.listMeters();
      const duration = performance.now() - start;

      const meters = Array.isArray(result) ? result : result.data || [];
      const count = meters.length;
      const meterNames = meters.slice(0, 3).map((m: { name: string }) => m.name).join(', ');

      return {
        name: 'Meters List',
        success: true,
        duration,
        message: `Found ${count} meter(s)`,
        details: count > 0 ? `Meters: ${meterNames}${count > 3 ? '...' : ''}` : undefined,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Meters List',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Meters List',
        success: false,
        duration,
        message: err.message || 'Failed to list meters',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/meters.py
"""Meter operation checks."""
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _list_meters_check(ctx: CheckContext) -> CheckResult:
    """List available meters."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'list_meters'):
            return CheckResult(
                name="meters_list",
                success=True,
                duration=0,
                message="Skipped (list_meters not available)",
                details="The list_meters method is not available in the SDK"
            )

        result = client.list_meters()

        if hasattr(result, 'data'):
            count = len(result.data)
            meters = [getattr(m, 'name', str(m)) for m in result.data[:3]]
        elif isinstance(result, list):
            count = len(result)
            meters = [getattr(m, 'name', str(m)) for m in result[:3]]
        else:
            count = 1
            meters = [str(result)]

        return CheckResult(
            name="meters_list",
            success=True,
            duration=0,
            message=f"Found {count} meter(s)",
            details=f"Meters: {', '.join(meters)}" if meters else None
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="meters_list",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="meters_list",
            success=False,
            duration=0,
            message=f"Failed to list meters: {e}"
        )


list_meters_check = Check(
    name="meters_list",
    description="List available meters",
    run=_list_meters_check
)
